File: Python3/1249. Minimum Remove to Make Valid Parentheses.py
# 100ms, 88.06%; 14.2MB, 100.00%
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        st, deleted = [], []
        for i, c in enumerate(s):
            if c == '(':
                st.append(i)
            elif c == ')':
                if st:
                    st.pop()
                else:
                    deleted.append(i)
        while st:
            deleted.append(st.pop())
        if not deleted: return s

        deleted.sort()
        deleted.append(len(s))
        res, i = [], 0
        for j in deleted:
            res.append(s[i:j])
            i = j + 1
        return ''.join(res)

File: 1201 - 1400/1249. Minimum Remove to Make Valid Parentheses.cc
// 24ms, 94.85%; 14.5MB, 100.00%
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string minRemoveToMakeValid(const string &s) {
        int slen = s.size();
        stack<int> st;
        vector<int> deleted;
        deleted.reserve(slen);
        for (int i = 0; i < slen; ++i) {
            if (s[i] == '(') {
                st.push(i);
            } else if (s[i] == ')') {
                if (!st.empty()) {
                    st.pop();
                } else {
                    deleted.push_back(i);
                }
            }
        }
        while (!st.empty()) {
            deleted.push_back(st.top());
            st.pop();
        }
        deleted.shrink_to_fit();
        if (deleted.empty()) return s;

        sort(deleted.begin(), deleted.end());
        deleted.push_back(slen);
        int i = 0;
        string res = "";
        for (auto j : deleted) {
            res += s.substr(i, j - i);
            i = j + 1;
        }
        return res;
    }
};
